### ISSUE TITLE
Iterative refinement in high dimensions

### ISSUE THREAD

**Issue by mforets** [JuliaReach/LazySets.jl#969]
Feature: overapproximate with an HPolytope in high dims, similar to what it's done for HPolygon. See for example [1].

[1] [The modified method of refined bounds for polyhedral approximation of convex polytopes](https://link.springer.com/article/10.1134/S0965542508060055), A. V. Lotov and A. I. Pospelov.

**Comment by maartenvd** [comment]
in #1691 you mention that this specific feature is probably infeasible. There are algorithms for approximating a polytope with a simpler one. Is it infeasible because we wouldn't know anything about the exact error?

Are there over/unerapproximations for a general VPolytope with a simpler one, in lazysets?

**Comment by schillic** [comment]
> Is it infeasible because we wouldn't know anything about the exact error?

It is infeasible in practice simply because you would need too many directions. Here I really assume a large number of dimensions.
(Plus, the implementation is nontrivial. This is the main reason why it is missing.)

> Are there over/unerapproximations for a general VPolytope with a simpler one, in lazysets?

Overapproximation: Indirectly. You can define a set of directions and obtain the corresponding template polytope with the facets with those normal directions. The result is an `HPolytope`, which you can convert to `VPolytope` afterward. But note that this may be very expensive in high dimensions. You usually want to stick to the `HPolytope`.

Underapproximation: Yes. It follows the same idea, but you get a `VPolytope` whose vertices are extreme points in the given directions.

You can either use `CustomDirections` to define the directions yourself or use any of the [predefined template directions](https://juliareach.github.io/LazySets.jl/dev/lib/approximations/template_directions/). You can use them like this (example for `BoxDirections`):

```julia
julia> X = # define your set here
julia> dirs = BoxDirections(dim(X))
julia> overapproximate(X, dirs)
julia> underapproximate(X, dirs)
```

In 3D, you can use `SphericalDirections`. But in higher dimensions, I think we do not have any functionality that gives you a simplified polytope for a given error bound.

**Comment by maartenvd** [comment]
I am well out of my depth here, but aren't the papers by löhne exactly about this problem, and avoid having to use this exponentially scaling number of directions? https://arxiv.org/pdf/2103.06613

**Comment by schillic** [comment]
I was not aware of this reference. I guess it depends on your definition of "closeness." From a quick scan, they achieve a Hausdorff distance that depends on the dimension. I guess this can be a useful result, but it is not what I had in mind above (I was thinking about a constant Hausdorff distance).

Are you interested in implementing this feature?

**Comment by maartenvd** [comment]
If I'm able to find some time I'll try implementing a toy version of that algorithm

**Comment by maartenvd** [comment]
After some tinkering, it looks like this is really just an abstracted version of kamenev's algorithm ... Extending kamenev's algorithm for higher dimensions (I think lazysets only has something in there for polygons?), with a simple-but-slow version as simple as

```
function outer_kamenev(v::VPolytope, eps)
    # get an initial simple bounding box
    outer = convert(VPolytope,overapproximate(v))
    
    inner_faces = constraints_list(v)
    while true
        errors = zeros(Float64,length(inner_faces))
        for (i,face) in enumerate(inner_faces)
            dir,rho_inner = face.a, face.b
            errors[i] = ρ(dir,outer)-rho_inner
        end

        (approximation_error, worstface) = findmax(errors)
        approximation_error < eps && return outer
        outer = convert(VPolytope,inner_faces[worstface]∩outer)
    end

    return outer
end
```

gives something that frequently crashes. I am frequently converting between H/V representation for outer, which cannot be stable, and is probably the issue.

**Comment by schillic** [comment]
~~What is `inner_faces` in your code?~~ Never mind, I am blind.

> I think lazysets only has something in there for polygons?

Exactly.

**Comment by schillic** [comment]
I assume that `inner_faces` is just a temporary placeholder for you, because otherwise I think your code will not do anything interesting. It will return the original polytope unless some constraints are contributing very little to the set size, right? Maybe that is what you wanted, but then there is a misunderstanding because that is not what this issue was about.

To explain the difference: the downside in your code is that `inner_faces` assumes that the input set is a polytope and you can construct the H-representation explicitly. So in a sense you have already spent the effort to construct the full H-representation. We do not have the use case to simplify a polytope which you have already represented explicitly. In LazySets, (a) the set is often not given explicitly but only as a lazy object, so computing the constraints would be really heavy, and (b) the set may not even be polytopic. You can use our 2D approximation for a Euclidean ball for instance and it will give you a polygonal approximation with the chosen error bound.

I also do not understand why you want `outer` to be in V-representation. I think it would be better in H-representation.

**Comment by maartenvd** [comment]
Hey, thanks for being patient!

Just to see if I understand, my code only works for polytopes, while this issue is meant to track general iterative refinement for all supported sets in lazysets? Which would require one that only ever calculates support vectors? And indeed, the toy algorithm I wrote down assumes willingness to convert to HRepresentation of the input polygon, which can scale very very badly.  That can definitely be avoided.

Though I'm confused why you think inner_faces is just a placeholder, this is the output of that algorithm when given a bunch of points on a circle as input set. It may not be a very good algorithm, but it does provide an outer approximation. I just now found a very related algorithm mentioned by kamenev.

<img width="664" height="444" alt="Image" src="https://github.com/user-attachments/assets/cd109b57-0b76-4722-acd8-313c12635d04" />

Even the linked paper at the start of this thread does not appear to describe an overapproximation, but rather an inner approximation.

**Comment by schillic** [comment]
Yes, sorry, what I meant was: `v` (the input) is a polytope, so you are starting with something that I would consider a valid output. Again, my original assumption was that you get an arbitrary (bounded) set as input. I agree that if you have a way to sample this set at the boundary sufficiently well, you can consider that as the input and then use your algorithm. So I would say it can still make sense to add this feature.

**Comment by maartenvd** [comment]
That makes sense! Is there a reason not to extend something like kamenev's algorithm to just higher dimensions, something like this? 

```

function kamenev(v::LazySet, eps)
    

    # v representation of an inner approximation
    inner_V = underapproximate(v,BoxDirections(dim(v)))
    inner_H = convert(HPolytope,inner_V) # H representation of the inner polytope

    # in this initial loop, we will construct the outer approximation
    spaces = eltype(constraints_list(inner_H))[]
    for face in constraints_list(inner_H)
        dir,rho_inner = face.a, face.b
        support_function = ρ(dir,v)
        push!(spaces,HalfSpace(dir,support_function))
    end


    outer = reduce(∩,spaces) # the outer approximation
    
    
    while true

        errors = Float64[]
        for face in constraints_list(inner_H)
            dir,rho_inner = face.a, face.b
            support_function = ρ(dir,outer)
            push!(errors,support_function-rho_inner )
        end

        (max_err,add_ind) = findmax(errors)
        
        max_err<eps && break
        worst_face = constraints_list(inner_H)[add_ind]
        
        push!(inner_V.vertices,σ(worst_face.a,v)) # this point must be added to the list of inner vertices
        inner_V = VPolytope(inner_V.vertices)
        inner_H = convert(HPolytope,inner_V) # this isn't ideal. inner_H was still largely correct

        outer = outer ∩ HalfSpace(worst_face.a,ρ(worst_face.a,v))       
    end
   
end
```

**Comment by schillic** [comment]
The first time the code is quite slow, probably because of external libraries being compiled, but it seems to work like a charm (when `return`ing `outer` in the end 🙂)! Thanks for sharing!
Note that the approximation is not symmetric, but maybe this is to be expected.
Do you want to submit a pull request? I would ask to call the function `overapproximate` and add a test case.

```julia
julia> using LazySets, GLMakie
julia> X = Ball2(zeros(3), 1.);  # unit sphere in 3D
julia> Y = kamenev(X, 1e-1);
julia> Z = Sphere(Point3f(0), 1f0);  # native Makie baseline for comparison

julia> mesh(Z, color = :blue);
julia> LazySets.plot3d!(Y, color=:yellow, alpha=0.5);
```

<img width="600" height="450" alt="Image" src="https://github.com/user-attachments/assets/4371df5a-3624-425d-970d-fc8796ad5e71" />

**Comment by maartenvd** [comment]
I can make it a bit nicer before creating a PR, but do you know if there's a more efficient way for updating the inner hpolytope? we only add a single point, so most h-planes should still be correct.

You also get an inner approximation, do you still want to call it overapproximate?

**Comment by schillic** [comment]
> I can make it a bit nicer before creating a PR, but do you know if there's a more efficient way for updating the inner hpolytope? we only add a single point, so most h-planes should still be correct.

No, I do not think there is a trick to exploit when adding a point. In 2D we sort the constraints, which allows better reasoning, but in higher dimensions, it is unclear how to sort. So you cannot know in advance which constraints need to be updated. You need to go through them all and check whether they contain the point.
But you do not have to worry about efficiency too much. This can always be improved later. Currently it is a strict improvement to have a new feature.

> You also get an inner approximation, do you still want to call it overapproximate?

I see. You can call the method something else (e.g., `kamenev`) and let it return both the outer and the inner approximation. This method should not be exported. And then add another `overapproximate` method that calls this method and only returns the outer approximation.
